cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,742 @@
+    case NC_STR: return " \"";
+    case NC_VAL: return " V";
+    case NC_FN: return " F";
+    case NC_ERR: return "ER";
+    default: return "UN";
+  };
+};
+auto to_int    = [](auto v) { return std::to_string(static_cast<int>(v)); };
+auto print_vec = [](auto const& cpu, auto const name, auto converter) {
+  for (auto const& v : cpu)
+    printf("%3s,", converter(v).c_str());
+  std::cout << name << std::endl;
+};
+
+void print_tree(host_span<SymbolT const> input,
+                tree_meta_t const& d_gpu_tree,
+                rmm::cuda_stream_view stream = cudf::default_stream_value)
+{
+  print_vec(cudf::detail::make_std_vector_async(d_gpu_tree.node_categories, stream),
+            "node_categories",
+            to_cat);
+  print_vec(cudf::detail::make_std_vector_async(d_gpu_tree.parent_node_ids, stream),
+            "parent_node_ids",
+            to_int);
+  print_vec(
+    cudf::detail::make_std_vector_async(d_gpu_tree.node_levels, stream), "node_levels", to_int);
+  auto node_range_begin = cudf::detail::make_std_vector_async(d_gpu_tree.node_range_begin, stream);
+  auto node_range_end   = cudf::detail::make_std_vector_async(d_gpu_tree.node_range_end, stream);
+  print_vec(node_range_begin, "node_range_begin", to_int);
+  print_vec(node_range_end, "node_range_end", to_int);
+  for (int i = 0; i < int(node_range_begin.size()); i++) {
+    printf("%3s ",
+           std::string(input.data() + node_range_begin[i], node_range_end[i] - node_range_begin[i])
+             .c_str());
+  }
+  printf(" (JSON)\n");
+}
+
+// input node tree, col_id, row_offset of nodes.
+// output column tree, col_id, max_row_offset of columns.
+std::tuple<tree_meta_t, rmm::device_uvector<NodeIndexT>, rmm::device_uvector<size_type>>
+reduce_to_column_tree(tree_meta_t& tree,
+                      device_span<NodeIndexT> col_ids,
+                      device_span<size_type> row_offsets,
+                      rmm::cuda_stream_view stream,
+                      rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  //   1. sort_by_key {col_id}, {row_offset} //stable?
+  rmm::device_uvector<NodeIndexT> node_ids(row_offsets.size(), stream);
+  thrust::sequence(rmm::exec_policy(stream), node_ids.begin(), node_ids.end());
+  thrust::stable_sort_by_key(rmm::exec_policy(stream),
+                             col_ids.begin(),
+                             col_ids.end(),
+                             thrust::make_zip_iterator(node_ids.begin(), row_offsets.begin()));
+  auto counting_it = thrust::make_counting_iterator<size_type>(0);
+  auto num_columns = thrust::count_if(rmm::exec_policy(stream),
+                                      counting_it,
+                                      counting_it + col_ids.size(),
+                                      [col_ids = col_ids.begin()] __device__(auto i) {
+                                        return i == 0 || col_ids[i] != col_ids[i - 1];
+                                      });
```

Comment:
https://github.com/NVIDIA/thrust/commit/8aecfe5e06962ec8fe32d307b821653523959190 🙃 
```suggestion
  auto num_columns = thrust::unique_count(rmm::exec_policy(stream), col_ids.begin(), col_ids.end());
```

Reply:
Thanks. This is why I prefer old thrust documentation format. All in one page.